### PR TITLE
feat(images): integrate image metadata stripping and enhance file handling

### LINKED ISSUE #149: [Feature]: Strip sensitive metadata from uploaded images
### Describe the problem

Uploaded images such as profile avatar and those attached to posts are not yet processed to remove sensitive metadata such as device used and geolocation.

### Describe the solution

Strip exif metadata before upload, see same done here: https://github.com/pubky/pubky-app/pull/1836
Also consider not using the same filename  - `name` in payload PUT

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Nirzhuk**
close #149 

- Added `stripImageMetadata` function to sanitize image files by removing sensitive metadata and obfuscating filenames.
- Updated `FileNormalizer.toFileAttachment` to utilize the new image sanitization process for JPEG, PNG, and WebP formats.
- Implemented tests for the `stripImageMetadata` function to ensure correct behavior for various image types and error handling.
- Enhanced existing tests in `file.normalizer.test.ts` to cover new functionality and edge cases related to image uploads.

Quite hard to give a demo but I will decribe how can be tested:

- Upload a Image normally
- Go to Network Request to check the name in the query it's a random name
- Go to Application indexedDb (franky)
- enter files_details and the files will have random names and values

### CODE REVIEW

**Comment by Taewa** [comment]
@greptile-apps
Review again plz

File: src/core/pipes/file/file.normalizer.test.ts [hunk 1]
```diff
@@ -11,8 +11,10 @@ import {
 import { AppError } from '@/libs/error/error';
 import { ValidationErrorCode } from '@/libs/error/error.codes';
 import { ErrorCategory, ErrorService } from '@/libs/error/error.types';
-import { FileNormalizer } from '@/pipes/file/file.normalizer';
-import { PubkySpecsSingleton } from '@/pipes/pipes.builder';
+import * as ImageSanitizer from '@/libs/image/stripImageMetadata';
+
```
[the remark this answers is not in the document]
Comment (Taewa) [reply src/core/pipes/file/file.normalizer.test.ts]: Let's keep our new style. 
`import * as ...` is not welcomed anymore unless there is a good reason to use.

You can use something like
```
import { stripImageMetadata } from '@/libs/image/stripImageMetadata';

// ...

vi.mock('@/libs/image/stripImageMetadata', () => ({
  stripImageMetadata: vi.fn(),
}));

// ...

vi.mocked(stripImageMetadata).mockImplementation(async (file: File) => file);
```

File: src/core/pipes/file/file.normalizer.test.ts [hunk 2]
```diff
@@ -115,6 +118,36 @@ describe('FileNormalizer', () => {
...
+        expect(ImageSanitizer.stripImageMetadata).toHaveBeenCalledWith(file);
+        expect(mockBuilder.createFile).toHaveBeenCalledWith(
+          'obfuscated123.jpg',
+          FILE_TEST_DATA.blobUrl,
+          'image/jpeg',
+          5000,
+        );
+      });
+
+      it('should keep original filename for non-image uploads', async () => {
```
Comment (Taewa) [inline src/core/pipes/file/file.normalizer.test.ts]: Just out of curiosity, we should keep original file name if it is not an image? Rr any kind of upload should be obfuscated?

File: src/core/pipes/file/file.normalizer.ts [hunk 3]
```diff
@@ -10,8 +12,9 @@ export class FileNormalizer {
 
   static async toFileAttachment({ file, pubky }: TUploadFileParams): Promise<TFileAttachmentResult> {
```
Comment (Taewa) [inline src/core/pipes/file/file.normalizer.ts]: A finding with AI


[1] high | no-io-in-pipes | src/core/pipes/file/file.normalizer.ts:15
What's wrong: FileNormalizer calls stripImageMetadata(file) from a pipe. That helper uses DOM APIs and side effects such as URL.createObjectURL, Image, canvas, and crypto.randomUUID.

Why it matters: Project rules require src/core/pipes/** to stay pure, with no IO or side effects.

Fix: Move image sanitization to an IO boundary such as Application/Service, and pass an already sanitized File or pure file metadata into FileNormalizer.

File: src/libs/image/stripImageMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+ * - All image types: replace original filename with an obfuscated one
+ */
+export async function stripImageMetadata(file: File): Promise<File> {
+  if (!file.type.startsWith('image/')) {
+    return file;
+  }
+
+  const obfuscatedName = generateObfuscatedImageFileName(file);
+
+  if (!MIME_TYPES_WITH_CANVAS_SANITIZATION.has(file.type)) {
```
Comment (Taewa) [inline src/libs/image/stripImageMetadata.ts]: A finding with AI. I am not sure what kind of personal sensitive metadata can be in metadata honestly. But I think it's worth to share with you.

[2] medium | security/privacy | src/libs/image/stripImageMetadata.ts:104
What's wrong: image/svg+xml is classified as an image, but its original bytes are uploaded unchanged.

Why it matters: Sensitive data inside SVGs, such as <metadata>, comments, or editor traces, is not removed. Filename obfuscation alone does not satisfy the “strip image metadata” requirement.

Fix: Treat SVG as a generic file, reject it for image uploads, or add a safe sanitization/rasterization path. Update the tests to assert the chosen safe behavior.

File: src/libs/image/stripImageMetadata.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * Removes sensitive metadata from supported image formats before upload.
+ * - JPEG/PNG/WebP: re-encodes via canvas to strip metadata (fail-closed on errors)
+ * - JPEG/WebP: encoded at quality 1.0 to avoid browser default quality reduction
+ * - GIF/SVG and other image types: keep bytes to avoid visual regressions
+ * - All image types: replace original filename with an obfuscated one
+ */
+export async function stripImageMetadata(file: File): Promise<File> {
+  if (!file.type.startsWith('image/')) {
```
Comment (Taewa) [inline src/libs/image/stripImageMetadata.ts]: A finding with AI

[3] medium | security/privacy | src/libs/image/stripImageMetadata.ts:98
What's wrong: Image detection relies only on the client-provided file.type.

Why it matters: Browser-provided MIME types can be missing or incorrect. A JPEG/PNG/WebP file with an empty or wrong MIME type can bypass EXIF stripping and filename obfuscation.

Fix: Detect supported image formats from magic bytes, or at least combine MIME type checks with a validated extension fallback. Add a test for an image file where file.type is empty.

**Review by Taewa — CHANGES_REQUESTED** [review]
I left a couple of feedbacks

File: src/core/pipes/file/file.normalizer.test.ts (hunk 2 above)
Comment (Nirzhuk) [reply src/core/pipes/file/file.normalizer.test.ts]: @Taewa we allow pdfs, so I imagine and we show the name in the attached part of the post, so showing random strings removes the confidence of that file to the user. 

My-normal-pdf.pdf
e18dg73j1h218w.pdf

**Comment by Nirzhuk** [comment]
@greptile-apps
Review again plz

**Comment by Taewa** [comment]
## Edied:
Ok, I was wrong. There is a security issues on that.

## Original:
Regarding SVG: The change doesn’t sanitize SVG internals beyond renaming. I’m fine approving without stripping XML-level metadata—I’d assume meaningful personal leakage there is uncommon unless we tighten requirements later.

File: src/libs/image/stripImageMetadata.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+ */
+export async function stripImageMetadata(file: File): Promise<File> {
+  const imageMimeType = await detectImageMimeType(file);
+  if (!imageMimeType) {
+    return file;
+  }
+
+  const obfuscatedName = generateObfuscatedImageFileName(file, imageMimeType);
+
+  if (!MIME_TYPES_WITH_CANVAS_SANITIZATION.has(imageMimeType)) {
```
Comment (Taewa) [inline src/libs/image/stripImageMetadata.ts:331]: A finding with AI.

[1] critical | Security | stripImageMetadata.ts:211

What's wrong: SVG files are passed through unchanged. The function is named stripImageMetadata but for image/svg+xml it only renames the file — <script>, on* event handlers, <foreignObject>, external <use
href>, etc. all survive.
Why it matters: This function is the upload gate (called from src/core/application/file/file.ts toFileAttachment). Any malicious SVG is stored as-is on the homeserver. If any consumer ever renders it
inline (dangerouslySetInnerHTML, <use href>, direct navigation to the blob URL), it becomes stored XSS.
The doc comment "keep bytes to avoid visual regressions" acknowledges the passthrough but offers no
security rationale.
Fix: Either (a) reject image/svg+xml outright and surface a validation error — simplest, matches the
fail-closed posture used for raster encode failures; or (b) sanitize with DOMPurify in SVG mode:
DOMPurify.sanitize(text, { USE_PROFILES: { svg: true, svgFilters: true } }) before wrapping in a new
File.

File: src/libs/image/stripImageMetadata.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    hasPrefix(bytes, [0x52, 0x49, 0x46, 0x46]) &&
+    bytes.length >= 12 &&
+    bytes[8] === 0x57 &&
+    bytes[9] === 0x45 &&
+    bytes[10] === 0x42 &&
+    bytes[11] === 0x50
+  );
+}
+
+function isSvgSignature(bytes: Uint8Array): boolean {
```
Comment (Taewa) [inline src/libs/image/stripImageMetadata.ts:96]: A finding with AI

[2] high | Security | stripImageMetadata.ts:79

What's wrong: isSvgSignature decodes the first 512 raw bytes as UTF-8 with a default TextDecoder (does
not throw on invalid sequences) and matches startsWith('<svg') / startsWith('<?xml') + includes('<svg').
A file with no Content-Type and no extension whose first bytes are ASCII <svg is promoted to
image/svg+xml.
Why it matters: Combined with finding [1], an attacker submits a file with type:'' and no extension;
magic-byte detection tags it image/svg+xml and the bytes ship unmodified. The other signature checks use
precise binary prefixes; the SVG check is the loosest by far.
Fix: Tighten the check (require <?xml or <svg followed by whitespace / > / attribute char; verify
well-formedness via DOMParser) and/or remove SVG from the auto-detect path so SVG is only honored when
file.type or the extension explicitly says so. Best combined with [1].

**Review by Taewa — CHANGES_REQUESTED** [review]
AI flagged this SVG passthrough as a potential stored XSS (SVG can contain <script> / on* handlers), plus 8 more findings. AI isn't always right, but file-handling code is worth double-checking — please run our `code-review` skill :)

**Comment by Nirzhuk** [comment]
@greptile-apps review it again please

**Comment by Taewa** [comment]
It renders SVG

https://github.com/user-attachments/assets/b8bcb8df-4d20-46a0-9db5-17a4430b3ba6

**Comment by Nirzhuk** [comment]
@Taewa https://github.com/pubky/pubky-app/issues/1861

**Review by Taewa — APPROVED** [review]
LGTM.

Note that there is a follow-up task
